book-2-platform-api/chapters/INTEGRATION_TESTS.md

```diff
@@ -0,0 +1,279 @@
+
+1. An HTTP client sends a `POST` request to the `/animal` route.
+1. The `Post()` method of the `AnimalController` is called.
+1. ADO.NET is used to insert the data into the corresponding Animal table in the database.
+1. The primary key is returned from the operation and the `Id` property of the `Animal` object instance is updated with that new value.
+1. The `CreatedAtRoute()` method is called in order the generate the new fully qualified URL of the resource that was just created (e.g. `http://localhost:5000/animal/8`).
+1. A response is generated with a 200 status code by the `CreatedAtRoute()` method and flung back at the client.
+
+The following test will mimic a client completing the create process and ensure that it works as expected.
+
+```cs
+[Fact]
+public async Task Test_Create_Animal()
+{
+    /*
+        Generate a new instance of an HttpClient that you can
+        use to generate HTTP requests to your API controllers.
+        The `using` keyword will automatically dispose of this
+        instance of HttpClient once your code is done executing.
+    */
+    using (var client = new APIClientProvider().Client)
+    {
+        /*
+            ARRANGE
+        */
+
+        // Construct a new student object to be sent to the API
+        Student jack = new Student
+        {
+            Name = "Jack",
+            Breed = "Cocker Spaniel",
+            Age = 4,
+            HasShots = true
+        };
+
+        // Serialize the C# object into a JSON string
+        var jackAsJSON = JsonConvert.SerializeObject(jack);
+
+
+        /*
+            ACT
+        */
+
+        // Use the client to send the request and store the response
+        var response = await client.PostAsync(
+            "/api/animals",
+            new StringContent(jackAsJSON, Encoding.UTF8, "application/json")
+        );
+
+        // Store the JSON body of the response
+        string responseBody = await response.Content.ReadAsStringAsync();
+
+        // Deserialize the JSON into an instance of Animal
+        var newJack = JsonConvert.DeserializeObject<Animal>(responseBody);
+
+
+        /*
+            ASSERT
+        */
+
+        Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+        Assert.Equal("Jack", newJack.Name);
+        Assert.Equal("Cocker Spaniel", newJack.Breed);
+        Assert.Equal(4, newJack.Age);
+    }
+}
+```
+
+## What is that ClientProvider?
+
+Since these tests are being created and executed by Visual Studio, you need a client in order to make an HTTP request. Normally this is done via Chrome or Postman - both clients for an API. ASP.NET and C# provides some classes that allow you to create a client that can perform HTTP requests.
+
+The following class should be added to your integration testing project for use in your testing classes.
+
+> APIClientProvider.cs
+
+```cs
+using Microsoft.AspNetCore.Mvc.Testing;
+using StudentExercisesAPI;
+using System.Net.Http;
+using Xunit;
+
+namespace TestStudentExercisesAPI
```

Comment:
Shouldn't have "StudentExercises" in the namespace. Maybe `TestAnimalAPI`?

Reply:
Yep, changed